### ISSUE TITLE
AsyncEngineDeadError when inference request aborted. Can I restart vllm-engine?

### ISSUE THREAD

**Issue by s-natsubori** [vllm-project/vllm#2475]
I trying use vllm with fastAPI as backend for [llm-vscode](https://github.com/huggingface/llm-vscode).
Client send many requests for server, and abandon some requests when user write code before return inference results.
on that case AsyncEngineDeadError happens frequently.

My question
- to avoid this, can I fix my code?
- or can restart vllm engine with some code?

## ENV
- docker image based nvcr.io/nvidia/pytorch:23.10-py3
- Nvidia Driver Version: 525.85.12
- CUDA Version: 12.0
- GPU Tesla T4
- Model octocoder(awq)
- transformers==4.36.0
- autoawq==0.1.8
- xformers==0.0.23.post1
- vllm build from latest main branch

## Engin Args
```
INFO 01-18 02:14:41 llm_engine.py:70] Initializing an LLM engine with config: model='/usr/local/model/llm', tokenizer='/usr/local/model/llm', tokenizer_mode=auto, revision=None, tokenizer_revision=None, trust_remote_code=True, dtype=torch.float16, max_seq_len=8192, download_dir=None, load_format=auto, tensor_parallel_size=2, quantization=awq, enforce_eager=True, seed=0)
```
## My Code(Partial)
``` python
    async def _get_vllm_res(self,request:Chat_Request
                            , inputs:BatchEncoding
                            , sampling_params:SamplingParams
                            , req_client: Request
                            , return_final:bool=False)->AsyncGenerator[str, None]:
        request_id = random_uuid()
        results_generator = self.engine.generate(
            prompt=None,
            prompt_token_ids=inputs.input_ids,
            sampling_params=sampling_params, 
            request_id=request_id
        )
        outputed_text = ""
        final_output = None
        async for request_output in results_generator:
            if req_client and await req_client.is_disconnected():
                # Abort the request if the client disconnects.
                info("abort")
                await self.engine.abort(request_id)
                return
            output = request_output.outputs[0].text
            yield output.replace(outputed_text, "")
            outputed_text = output
```
## Error log
```
INFO 01-18 02:17:55 llm_engine.py:823] Avg prompt throughput: 1287.5 tokens/s, Avg generation throughput: 0.0 tokens/s, Running: 1 reqs, Swapped: 0 reqs, Pending: 0 reqs, GPU KV cache usage: 0.5%, CPU KV cache usage: 0.0%
INFO:     10.134.64.166:56550 - "POST /code_generate HTTP/1.1" 200 OK
INFO 01-18 02:18:01 llm_engine.py:823] Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 13.8 tokens/s, Running: 0 reqs, Swapped: 0 reqs, Pending: 0 reqs, GPU KV cache usage: 0.4%, CPU KV cache usage: 0.0%
INFO:     10.134.64.166:56550 - "POST /code_generate HTTP/1.1" 200 OK
INFO 01-18 02:18:06 llm_engine.py:823] Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 29.2 tokens/s, Running: 1 reqs, Swapped: 0 reqs, Pending: 0 reqs, GPU KV cache usage: 0.5%, CPU KV cache usage: 0.0%
INFO:     10.134.64.166:56550 - "POST /code_generate HTTP/1.1" 200 OK
INFO 01-18 02:18:11 llm_engine.py:823] Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 17.2 tokens/s, Running: 1 reqs, Swapped: 0 reqs, Pending: 0 reqs, GPU KV cache usage: 0.5%, CPU KV cache usage: 0.0%
INFO:     10.134.64.166:56550 - "POST /code_generate HTTP/1.1" 200 OK
INFO 01-18 02:18:16 llm_engine.py:823] Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 17.1 tokens/s, Running: 1 reqs, Swapped: 0 reqs, Pending: 0 reqs, GPU KV cache usage: 0.5%, CPU KV cache usage: 0.0%
INFO:     10.134.64.166:56550 - "POST /code_generate HTTP/1.1" 200 OK
[2024-01-18 02:18:20,691] [INFO] /usr/local/api/chat_models/chat_local_vllm.py(161):_get_vllm_res: abort
INFO 01-18 02:18:22 llm_engine.py:823] Avg prompt throughput: 415.2 tokens/s, Avg generation throughput: 11.8 tokens/s, Running: 1 reqs, Swapped: 0 reqs, Pending: 0 reqs, GPU KV cache usage: 0.5%, CPU KV cache usage: 0.0%
INFO:     10.134.64.166:56551 - "POST /code_generate HTTP/1.1" 200 OK
INFO 01-18 02:18:28 llm_engine.py:823] Avg prompt throughput: 0.0 tokens/s, Avg generation throughput: 8.4 tokens/s, Running: 1 reqs, Swapped: 0 reqs, Pending: 0 reqs, GPU KV cache usage: 0.5%, CPU KV cache usage: 0.0%
[2024-01-18 02:18:28,030] [INFO] /usr/local/api/chat_models/chat_local_vllm.py(161):_get_vllm_res: abort
[2024-01-18 02:18:30,627] [INFO] /usr/local/api/chat_models/chat_local_vllm.py(161):_get_vllm_res: abort
[2024-01-18 02:18:30,629] [ERRO] /usr/lib/python3.10/asyncio/runners.py(44):run: Exception in callback functools.partial(<function _raise_exception_on_finish at 0x7f9d40834670>, request_tracker=<vllm.engine.async_llm_engine.RequestTracker object at 0x7f9d1da06fb0>)
handle: <Handle functools.partial(<function _raise_exception_on_finish at 0x7f9d40834670>, request_tracker=<vllm.engine.async_llm_engine.RequestTracker object at 0x7f9d1da06fb0>)>
Traceback (most recent call last):
  File "/usr/local/setup/vllm/vllm/engine/async_llm_engine.py", line 28, in _raise_exception_on_finish
    task.result()
  File "/usr/local/setup/vllm/vllm/engine/async_llm_engine.py", line 364, in run_engine_loop
    has_requests_in_progress = await self.engine_step()
  File "/usr/local/setup/vllm/vllm/engine/async_llm_engine.py", line 343, in engine_step
    request_outputs = await self.engine.step_async()
  File "/usr/local/setup/vllm/vllm/engine/async_llm_engine.py", line 190, in step_async
    all_outputs = await self._run_workers_async(
  File "/usr/local/setup/vllm/vllm/engine/async_llm_engine.py", line 231, in _run_workers_async
    all_outputs = await asyncio.gather(*coros)
  File "/usr/lib/python3.10/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/usr/local/lib/python3.10/dist-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "/usr/local/setup/vllm/vllm/worker/worker.py", line 197, in execute_model
    output = self.model_runner.execute_model(seq_group_metadata_list,
  File "/usr/local/lib/python3.10/dist-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "/usr/local/setup/vllm/vllm/worker/model_runner.py", line 524, in execute_model
    self.prepare_input_tensors(seq_group_metadata_list))
  File "/usr/local/setup/vllm/vllm/worker/model_runner.py", line 389, in prepare_input_tensors
    ) = self._prepare_decode(seq_group_metadata_list)
  File "/usr/local/setup/vllm/vllm/worker/model_runner.py", line 234, in _prepare_decode
    max_context_len = max(context_lens)
ValueError: max() arg is an empty sequence

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "uvloop/cbhandles.pyx", line 63, in uvloop.loop.Handle._run
  File "/usr/local/setup/vllm/vllm/engine/async_llm_engine.py", line 37, in _raise_exception_on_finish
    raise exc
  File "/usr/local/setup/vllm/vllm/engine/async_llm_engine.py", line 32, in _raise_exception_on_finish
    raise AsyncEngineDeadError(
vllm.engine.async_llm_engine.AsyncEngineDeadError: Task finished unexpectedly. This should never happen! Please open an issue on Github. See stack trace above for the actual cause.
```

**Comment by s-natsubori** [comment]
fixed with #2463

**Comment by s-natsubori** [comment]
fixed with #2463
